src/generic/utf8_zwegner_algorithm.h

```diff
@@ -0,0 +1,201 @@
+    // happen, two of these bit sequences would have to overlap. If
+    // this is the case, there is a leader byte before the second set
+    // of required continuation bytes (and thus before the bit that
+    // will be cleared by a carry). This leader byte will not be
+    // in the continuation mask, despite being required. QEDish.
+    // Which bytes are required to be continuation bytes
+    vmask2_t cont_required = this->last_cont;
+
+    // 2-byte lead: 1111____
+    const vmask_t bit_6 = bytes.shl<1>().high_bits_to_bitmask();
+    const vmask_t lead_2_plus = bit_7 & bit_6;       // 11______
+    cont_required += vmask2_t(lead_2_plus) << 1;
+
+    // 3-byte lead: 1111____
+    const vmask_t bit_5 = bytes.shl<2>().high_bits_to_bitmask();
+    const vmask_t lead_3_plus = lead_2_plus & bit_5; // 111_____
+    cont_required += vmask2_t(lead_3_plus) << 2;
+
+    // 4-byte lead: 1111____
+    const vmask_t bit_4 = bytes.shl<3>().high_bits_to_bitmask();
+    const vmask_t lead_4_plus = lead_3_plus & bit_4;
+    cont_required += vmask2_t(lead_4_plus) << 3;
+
+    const vmask_t cont = bit_7 ^ lead_2_plus;        // 10______ TODO &~ bit_6 might be fine, and involve less data dependency
+
+    // Check that continuation bytes match. We must cast req from vmask2_t
+    // (which holds the carry mask in the upper half) to vmask_t, which
+    // zeroes out the upper bits
+    //
+    // NOTE (@jkeiser): I turned the if() statement here into this->has_error for performance in
+    // success cases: instead of spending time testing the result and introducing a branch (which
+    // can affect performance even if it's easily predictable), we test once at the end.
+    // The ^ is equivalent to !=, however, leaving a 1 where the bits are different and 0 where they
+    // are the same.
+    this->cont_error |= cont ^ vmask_t(cont_required);
+
+    // Look up error masks for three consecutive nibbles. We need to
+    // AND with 0x0F for each one, because vpshufb has the neat
+    // "feature" that negative values in an index byte will result in 
+    // a zero.
+    //
+    // NOTE (@jkeiser): I removed the & 0x0F here because shr<4> already does that.
+    vec_t nibble_1_error = shifted_bytes.shr<4>().lookup_16<uint8_t>(
+        0, 0, 0, 0,
+        0, 0, 0, 0,
+        0, 0, 0, 0,
+
+        OVERLONG_2,   // [1100]000_         ________        Could have been encoded in 1 byte
+        0,
+        OVERLONG_3 |  // [1110]0000         100_____        Could have been encoded in 2 bytes
+          SURROGATE,  // [1110]1010         101_____        Surrogate pairs
+        OVERLONG_4 |  // [1111]0000         1000____        Could have been encoded in 3 bytes
+          TOO_LARGE | // [1111]0100         (1001|101_)____ > U+10FFFF
+          TOO_LARGE_2 // [1111](0101..1111) ________        > U+10FFFF
+    );
+
+    // NOTE (@jkeiser): I removed the & 0x0F here because lookup_lower_4() ignores the upper nibble
```

Comment:
`lookup_lower_4` seems to map straight to `pshufb` ?
`PSHUFB` does not ignore the upper nibble.

Reply:
Thanks for this; if I'm reading this right, it will ignore bits 4-6, and use bit 7 as a "set this one to zero instead of looking it up" bit, is that reading right?

There are several places we use lower_4_bits, one (IIRC) in `master` and one in my algorithm. Given that we're using it extensively in UTF-8 validation, I'm surprised it's passing the tests: specifically, the things we're looking up in that table *do* have high bits, and we *do* need them to yield nonzero values.

The two lookup_16 calls can safely omit the `& 0x0F`, because they are indeed already < 16.

Reply:
And wow, does that `&` have a massive performance impact. On the order of 10% degradation per `&`. Our whitespace and structural lookup uses `lookup_lower_4`, but it's possible those were already designed with the high bit in mind when I translated them. I'll make up a patch to `master` once I've figured out if there is an error in master; and I'll see if I can't make a better test.

It is pretty applicable here though. Literally all of the bit patterns we detect as errors start with a `1` in the high bit. If we flipped that bit we'd essentially get a 5 bit table lookup ("take a look at bits 3-6, but set the result to zero if the high bit is 0").

Reply:
the whitespace detection is correct in `master`

Reply:
Yeah, something doesn't add up right now. That code that you linked, which does not `& 0x0F`, is indeed detecting that C0 80 is an invalid sequence, even though the second nibble (the "0" in "C0") should have yielded 0 because there is a `1` in bit 7.